presto-main/src/main/java/io/prestosql/operator/HttpPageBufferClient.java

```diff
@@ -593,7 +619,15 @@ public PagesResponse handle(Request request, Response response)
                 boolean complete = getComplete(response);
 
                 try (SliceInput input = new InputStreamSliceInput(response.getInputStream())) {
+                    int magic = input.readInt();
+                    if (magic != SERIALIZED_PAGES_MAGIC) {
+                        throw new IllegalStateException(format("Invalid stream header, expected 0x%08x, but was 0x%08x", SERIALIZED_PAGES_MAGIC, magic));
+                    }
+                    long checksum = input.readLong();
+                    int pagesCount = input.readInt();
                     List<SerializedPage> pages = ImmutableList.copyOf(readSerializedPages(input));
+                    verifyChecksum(request.getUri(), checksum, pages);
+                    checkState(pages.size() == pagesCount, "Wrong number of pages, expected %s, but read %s", pagesCount, pages.size());
```

Comment:
How does this surface to the user? It'd be ideal if it could be mapped to a proper error code.

Reply:
replied here: https://github.com/prestosql/presto/pull/3438#issuecomment-617750235

Reply:
In practise, this is reachable only when checksum verification is OFF. This is then treated as if IO error occurred and retried.